### PR TITLE
Event database benchmark

### LINKED ISSUE #2480: Event database benchmark

### PR DESCRIPTION
**Pull request by Sriep**
## Changes
https://github.com/0chain/0chain/blob/662f7e9f1c03969840974a612868abb7bbb06dbf/code/go/0chain.net/smartcontract/benchmark/main/readme.md#event-database-benchmark
```
Event database benchmark
To run a event database benchmark set the option.event_database_benchmarks: true in the config file. This will test the events saved in the file indicated in the config file, option.event_database_event_file.

The event_database_event_file is a JSON file that contains a map of [event obejcts](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/dbs/event/event.go#L12). The map key is used as a label for displaying the results.

The use can make their own event_database_event_file file or use the one generated output from the smart-contract benchmarks. To do this set the two JSON files to be the same.

The option.smart_contract_event_file option will save the events generated by the smart-contract to a JSON file with the correct format to be used for the event database benchmark.

Example config file settings to link the smart-contract benchmark events to the event database benchmarks.

options:
  event_database_benchmarks: true
  smart_contract_event_file: "edb_in.json"
  event_database_event_file: "edb_in.json"
```
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | sprint-june-5  |
| gosdk  | staging  |
| zboxcli  | sprint-june-5  |
| zwalletcli  | sprint-june-5  |
| 0chain  | sprint-june-5  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | sprint-june-5  |
| gosdk  | buff-download  |
| zboxcli  | sprint-june-5  |
| zwalletcli  | sprint-june-5  |
| 0chain  | sprint-june-5  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

File: code/go/0chain.net/smartcontract/benchmark/main/cmd/suites.go [hunk 1]
```diff
@@ -219,5 +268,111 @@ func runSuite(
...
+}
+
+func runEventDatabaseSuite(
+	suite benchmark.TestSuite,
+	edb *event.EventDb,
+) []benchmarkResults {
+	var benchmarkResult []benchmarkResults
+	config.InitConfigurationGlobal(
+		edb.Config().Host,
+		"piers' port",
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/benchmark/main/cmd/suites.go]: please clean up

File: code/go/0chain.net/smartcontract/benchmark/main/config/benchmark.yaml [hunk 2]
```diff
@@ -1,12 +1,12 @@
 simulation:
-  num_clients: 210
-  num_active_clients: 210
-  num_miners: 100
-  num_active_miners: 100
-  nun_sharders: 20
-  nun_active_sharders: 20
-  num_allocations: 300
-  num_blobbers: 100
+  num_clients: 21
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/benchmark/main/config/benchmark.yaml]: rollback changes please

File: code/go/0chain.net/smartcontract/dbs/event/process.go [hunk 3]
```diff
@@ -261,12 +286,16 @@ func (edb *EventDb) addEventsWorker(ctx context.Context) {
...
 
-func (edb *EventDb) work(ctx context.Context,
-	gs *Snapshot, es blockEvents, currentPartition *int64) (*Snapshot, error) {
+func (edb *EventDb) Work(ctx context.Context,
+	gs *Snapshot, es BlockEvents, currentPartition *int64) (*Snapshot, error) {
 	tx := es.tx
 
 	var commit bool
 	defer func() {
+		if r := recover(); r != nil {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/process.go]: we recover in processEvent it should be enough

File: code/go/0chain.net/smartcontract/benchmark/main/cmd/options.go [hunk 4]
```diff
@@ -12,6 +24,28 @@ import (
...
+	bk.Vesting:         vestingsc.BenchmarkTests,
+	bk.VestingRest:     vestingsc.BenchmarkRestTests,
+	bk.MultiSig:        multisigsc.BenchmarkTests,
+	bk.ZCNSCBridge:     zcnsc.BenchmarkTests,
+	bk.ZCNSCBridgeRest: zcnsc.BenchmarkRestTests,
+	bk.Control:         control.BenchmarkTests,
+}
+
+func readEventDbTests(filename string) (map[string][]event.Event, error) {
+	file, _ := ioutil.ReadFile(filename)
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/benchmark/main/cmd/options.go:43]: why do we have file here, we need to execute events from transaction execution

**Review by dabasov — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/benchmark/main/config/benchmark.yaml (hunk 2 above)
Comment (Sriep) [reply code/go/0chain.net/smartcontract/benchmark/main/config/benchmark.yaml]: These changes are there for testing. It saves about 5-10 minutes each time I run the code. I'll change back when I have finished testing. These are changes I always make whenever I work on the benchmark code, and then change back when finished.

File: code/go/0chain.net/smartcontract/benchmark/main/cmd/options.go (hunk 4 above)
Comment (Sriep) [reply code/go/0chain.net/smartcontract/benchmark/main/cmd/options.go:43]: It adds flexibility at no extra cost. 

We should look to soft coding the smart contract benchmark tests similarly; however, that would mean writing common smart contract execute code.

File: code/go/0chain.net/smartcontract/benchmark/main/cmd/suites.go [hunk 5]
```diff
@@ -96,14 +113,38 @@ func runSuites(
 		}(suite, &wg)
 	}
 	wg.Wait()
+	data.EventDb.Close()
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/benchmark/main/cmd/suites.go:117]: should we close in defer()

Comment (Sriep) [reply code/go/0chain.net/smartcontract/benchmark/main/cmd/suites.go:117]: Creating a database with a template tends to fail if there are other open connections to the template database. Hence the timing of data.EventDb.Close() .
https://github.com/0chain/0chain/blob/cac33a11848d45799c3957d9c6005456c2a5aff5/code/go/0chain.net/smartcontract/dbs/postgresql/db.go#L45

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  |   |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  |   |
| zwalletcli  |   |
| 0chain  | edb_benchmark  |
| blobber  |   |
| authorizer  |   |
| 0box  |   |
| 0dns  |   |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  |   |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  |   |
| zwalletcli  |   |
| 0chain  | sprint-july-3  |
| blobber  |   |
| authorizer  |   |
| 0box  |   |
| 0dns  |   |

File: code/go/0chain.net/smartcontract/dbs/event/process.go [hunk 6]
```diff
@@ -264,74 +284,106 @@ func (edb *EventDb) addEventsWorker(ctx context.Context) {
...
+}
 
-	// process snapshot for none adding block events only
-	if isNotAddBlockEvent(es) {
-		gs, err = updateSnapshots(gs, es, tx)
+func (edb *EventDb) WorkAggregates(
+	gSnapshot *Snapshot,
+	blockEvents BlockEvents,
+) (*Snapshot, error) {
+	var err error
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/process.go]: return this error

File: code/go/0chain.net/smartcontract/dbs/postgresql/db.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	"fmt"
+
+	"0chain.net/chaincore/config"
+	"0chain.net/smartcontract/dbs"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+	"gorm.io/gorm/logger"
+)
+
+type PostgresDB struct {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/postgresql/db.go:13]: we already have script for postgres db

**Review by dabasov — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/dbs/postgresql/db.go (hunk 7 above)
Comment (Sriep) [reply code/go/0chain.net/smartcontract/dbs/postgresql/db.go:13]: This creates a link to the parent database. So here, `postgres` or PostgresDB is a particular PostgreSQL database. We only want it for one purpose, so it does not need the extra structure of an event database.

We can't clone a database with a template; say `event_db` if there are any active connections to the template database. In particular, we can't use a connection to a database to clone itself. We have to use another database; hence we open a connection to `postgres` the 'parent' database. 

I put the code in the `postgresql` package as it's a PostgreSQL-specific solution. 

It's the only way I could get the whole thing to work, but welcome other suggestions.

File: code/go/0chain.net/smartcontract/dbs/event/process.go (hunk 6 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/process.go]: @Sriep how is it resolved, I don't see you return this error anywhere

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**